src/util/hb_arena.c

```diff
@@ -5,58 +5,149 @@
 #include "../include/util/hb_arena.h"
 
 #include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
 #include <sys/mman.h>
 
 #define KB(kb) (1024 * kb)
 #define MB(mb) (1024 * KB(mb))
+#define HB_ARENA_PAGE_SIZE_DEFAULT KB(4)
 
-bool hb_arena_init(hb_arena_T* allocator, size_t size) {
-  allocator->memory = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+bool hb_arena_init(hb_arena_T* allocator, size_t initial_size) {
+  if (initial_size == 0) { initial_size = HB_ARENA_PAGE_SIZE_DEFAULT; }
 
-  if (allocator->memory == MAP_FAILED) {
-    allocator->memory = NULL;
-    allocator->position = 0;
-    allocator->capacity = 0;
+  char* memory = mmap(NULL, initial_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
 
-    return false;
-  }
+  if (memory == MAP_FAILED) { return false; }
+
+  allocator->head =
+    (hb_arena_page_T) { .next = NULL, .start = memory, .current = memory, .end = memory + initial_size };
 
-  allocator->position = 0;
-  allocator->capacity = size;
+  allocator->tail = &allocator->head;
+  allocator->default_page_size = initial_size;
 
   return true;
 }
 
 void* hb_arena_alloc(hb_arena_T* allocator, size_t size) {
-  size_t required_size = ((size + 7) / 8) * 8; // rounds up to the next 8 bytes
+  if (size == 0) { return NULL; }
+
+  size_t required_size = ((size + 7) / 8) * 8;
+
+  char* result = allocator->tail->current;
+
+  uintptr_t current = (uintptr_t) allocator->tail->current;
+  uintptr_t delta = 0;
+
+  if (current % 8 != 0) { delta = 8 - (current % 8); }
+
+  result += delta;
+
+  char* next = result + required_size;
+
+  if (next > allocator->tail->end) {
+    size_t next_page_size = allocator->default_page_size;
+
+    if (required_size > next_page_size) { next_page_size = required_size; }
```

Comment:
Suggestion: Go bigger than the required size.

If you end up having to make the page bigger because of the required size, it may make sense to already allocate some more, so that the next alloc doesn't have to allocate yet another page.

Reply:
What would the algorithm be in such a case? Just double the required size?

Reply:
For the bufferes we used twice the size, but this could fire back if you have large buffers, maybe use 1.5